### PR TITLE
Demonstrate what more robust storage might look like

### PR DESCRIPTION
**Pull request by mhammond**
This is a POC for #11810 and https://github.com/mozilla/application-services/issues/4872. There is a companion PR for Fenix which I'll link.

The point is that while we have fixed the most common manifestation of #11810  in app-services, there *will* still be cases with the same symptoms which we can't fix via code - particularly when there really is a corrupt sqlite database or no space on the device. We should try and avoid crashing in those cases and instead offer a degraded autofill experience.

This PR is just for autofill - it (a) tries to handle the exception opening the DB and (b) simulates such an exception just to see what happens. With this, Fenix doesn't crash, even when browsing to autofill test pages. It probably does crash if you try and explicitly add a creditcard, but that would (a) be easily fixed and (b) far far better than crashing as the app starts.

The question: is a strategy like this worth pursuing? I think each component would be slightly different but of a similar shape (ie, over time, we'd need to do something similar for all our components)

In https://github.com/mozilla/application-services/issues/4872, Ben floated the idea of trying to work out how to do this in rust and perform the same kind of short-circuiting there, but I pushed back a little on that - in particular, consider, say, `CreditCard add_credit_card(UpdatableCreditCardFields cc);` - I think the only way to sanely do this at a lower level would be to have that function return `null` in these error cases, which seems like a bad outcome and still likely to crash naive clients - I think it's better to let the app itself deal with catastrophic errors than having the lower-level components try and pretend everything is OK.

WDYT? cc @kbrosnan, @grigoryk, @gabrielluong, @bendk, @lougeniaC64

### CODE REVIEW

**Comment by gabrielluong** [comment]
- Would the expectation be that we will have to wrap all the getter, adder, updater methods around a try catch to avoid an exception if there is no storage? 
- Seeing the `hasStorage()` method and not seeing any usage of it, I am wondering when you might expect us to call it?
- From a product standpoint, what would be flow or messaging if someone was trying to add a credit card, but the storage isn't initialized because it's full?

**Comment by lougeniaC64** [comment]
TL;DR 👍🏾

I like how this solution isn't prescriptive and would provide consumers with a way to detect whether storage is present and render a degraded experience if it isn't. Deciding what happens when these storage failures occur should involve product direction as that degraded flow will impact the user and how they perceive the product's performance. I'll admit I'm not sure what a degraded user experience should be for autofill (is it in essence disabling it? resetting it?), but I'm more than certain that whatever it is will be better than simply crashing from the users' perspective.

**Comment by bendk** [comment]
Also :+1: from me.  This seems like a step in the right direction.  Also agreed on the `add_credit_card()` example.  I don't think we should try to handle things at the Rust level.

It seems to me that this is highly related to another question: who's responsible for what?  What code is application-services solely responsible for?  What code is the Fenix/iOS solely responsible for?  What code do we share the responsibility?

I'd like to propose that we formalize a pattern that I've noticed in the Fenix code:

  - There's a lower layer that app-services is solely responsible for.
     - This would be `AutofillCreditCardsAddressesStorage` and all the Rust code that sits underneath it
     - This layer should have a class that you can construct, even in the face of storage errors.  But in that case, most/all of it's methods will throw.
  - There's a middle layer that we share responsibility for.
     - In general, I think it's mostly delegates, adapters, etc.
     - This layer should present an interface that works even in the face of storage errors
     - For this particular issue, I think we should update `GeckoCreditCardsAddressesStorageDelegate` to check `hasStorage()`.  If that's false, then `onAddressesFetch()` would always return an empty list and `onAddressSave()` would be a no-op, etc.
  - There's a top layer that the Fenix/iOS team is soley responsible for.
    - You all know better than me what's in here :)
    - If product decides that there should be a UI for storage errors, then it would mostly be up to the Andorid/Fenix teams to implement it.  If it also required changes to the middle or lower layer, then our teams could work together to make those changes.

**Comment by mhammond** [comment]
> * Would the expectation be that we will have to wrap all the getter, adder, updater methods around a try catch to avoid an exception if there is no storage?

The expectation is more that the app itself should call `hasStorage()` *or* catch an exception. In my example above, consider:

    CreditCard add_credit_card(UpdatableCreditCardFields cc);

There's no way you can get back a `CreditCard` object in this scenario, and wrapping just this call in an exception handler probably doesn't make much sense - what might make more sense is that the app itself, in the layer above this, calls `hasStorage()` before it even offers the capability of adding a credit-card.

>     * Seeing the `hasStorage()` method and not seeing any usage of it, I am wondering when you might expect us to call it?

The callers would be in Fenix - see https://github.com/mhammond/fenix/commit/44c90ea1c69a079a462dde8febd600516c909470

>     * From a product standpoint, what would be flow or messaging if someone was trying to add a credit card, but the storage isn't initialized because it's full? What would be the "fix" to allow users to have a working storage and their associated features again after they encounter this messaging?

TBH, I think no messaging is OK - a "degraded experience". Keep in mind that the current behaviour is to crash on startup. The fact that autofill simply fails to work in a silent and possibly confusing way is still a massive improvement IMO. Obviously extra messaging could be done, but I think it's fine to do nothing.

An analogy is desktop Firefox when places.sqlite is corrupt - things might not work in slightly confusing ways, but there's no messaging and no crashing. Unlike Desktop, the mobile Firefoxes have the benefit that uninstalling and reinstalling the app will actually fix the problem (as all files are deleted on uninstall, whereas desktop leaves the profile behind on uninstall by default)

**Comment by mhammond** [comment]
Ben writes:
> I'd like to propose that we formalize a pattern that I've noticed in the Fenix code:

I agree 100% with this in general, although I wouldn't go as far as "formalize" - I think it depends on various factors. For example, with autofill in particular, the interactions with Gecko might make formalizing this tricky. Something like "usual pattern" or something slightly softer than implied by "formalize" :)

Lougenia writes:
>  I'm not sure what a degraded user experience should be for autofill

Exactly, which is why I'm reluctant to make it too formal - I think something like "don't crash, try to not mislead the user" is the yardstick, but anything more concrete than that depends on the component.
